File: backend/src/models/CartItem.js
const mongoose = require('mongoose');

const cartItemSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    cakeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Cake', required: true },
    quantity: { type: Number, required: true, default: 1, min: 1 },
    selectedSize: { type: String, required: true },
    selectedFlavor: { type: String, required: true },
    customMessage: { type: String, default: null },
    unitPrice: { type: Number, required: true },
  },
  { timestamps: true }
);

cartItemSchema.methods.toPublicJSON = function () {
  const cakeRef = this.cakeId;
  const cakeId =
    cakeRef && typeof cakeRef === 'object' && cakeRef._id
      ? cakeRef._id.toString()
      : cakeRef?.toString?.() ?? null;

  return {
    id: this._id.toString(),
    cakeId,
    quantity: this.quantity,
    selectedSize: this.selectedSize,
    selectedFlavor: this.selectedFlavor,
    customMessage: this.customMessage,
    unitPrice: this.unitPrice,
    createdAt: this.createdAt?.getTime(),
  };
};

module.exports = mongoose.model('CartItem', cartItemSchema);

File: flutter-app/cake_shop/lib/models/cart_item.dart
class CartItem {
  final String id;
  final String cakeId;
  final int quantity;
  final String selectedSize;
  final String selectedFlavor;
  final String? customMessage;
  final double unitPrice;
  String? cakeName;

  CartItem({
    required this.id,
    required this.cakeId,
    required this.quantity,
    required this.selectedSize,
    required this.selectedFlavor,
    this.customMessage,
    required this.unitPrice,
    this.cakeName,
  });

  double get lineTotal => unitPrice * quantity;

  factory CartItem.fromJson(Map<String, dynamic> json) {
    return CartItem(
      id: json['id'] as String,
      cakeId: json['cakeId'] as String,
      quantity: json['quantity'] as int,
      selectedSize: json['selectedSize'] as String,
      selectedFlavor: json['selectedFlavor'] as String,
      customMessage: json['customMessage'] as String?,
      unitPrice: (json['unitPrice'] as num).toDouble(),
    );
  }
}

class CartResponse {
  final List<CartItem> items;
  final double total;

  CartResponse({required this.items, required this.total});

  factory CartResponse.fromJson(Map<String, dynamic> json) {
    final items = (json['items'] as List)
        .map((e) => CartItem.fromJson(e as Map<String, dynamic>))
        .toList();
    return CartResponse(
      items: items,
      total: (json['total'] as num).toDouble(),
    );
  }
}
